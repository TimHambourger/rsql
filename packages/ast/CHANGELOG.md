# v1.0.0 (Sun May 10 2020)

#### üí• Breaking Change

- 1.0.0 release [#6](https://github.com/piotr-oles/rsql/pull/6) ([@piotr-oles](https://github.com/piotr-oles))
- feat: improved API and documentation + removed definitions ([@piotr-oles](https://github.com/piotr-oles))

#### üöÄ Enhancement

- feat: add support for custom comparison operators ([@piotr-oles](https://github.com/piotr-oles))
- feat: general improvements and documentation ([@piotr-oles](https://github.com/piotr-oles))
- feat: implement initial version of the RSQL parser ([@piotr-oles](https://github.com/piotr-oles))

#### üêõ Bug Fix

- docs: remove in-development warning ([@piotr-oles](https://github.com/piotr-oles))
- fix: make definition files compatible with IDEs [#5](https://github.com/piotr-oles/rsql/pull/5) ([@piotr-oles](https://github.com/piotr-oles))
- fix: make definition files compatible with IDEs ([@piotr-oles](https://github.com/piotr-oles))
- chore: use rollup to decrease build size ([@piotr-oles](https://github.com/piotr-oles))
- docs: update README.md ([@piotr-oles](https://github.com/piotr-oles))
- General improvements [#3](https://github.com/piotr-oles/rsql/pull/3) ([@piotr-oles](https://github.com/piotr-oles))
- chore: add registry url to each package.json ([@piotr-oles](https://github.com/piotr-oles))
- chore: add access 'public' entry to package.json ([@piotr-oles](https://github.com/piotr-oles))

#### Authors: 1

- Piotr Ole≈õ ([@piotr-oles](https://github.com/piotr-oles))