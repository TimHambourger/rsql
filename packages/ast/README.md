# @rsql/ast

Abstract Syntax Tree definitions and functions for RSQL

⚠️ WARNING: This package is still in development - API may break compatibility without upgrading major version! ⚠️

[![npm](https://img.shields.io/npm/v/@rsql/ast)](https://www.npmjs.com/package/@rsql/ast)
[![lerna](https://img.shields.io/badge/maintained%20with-lerna-cc00ff.svg)](https://lerna.js.org/)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier)
[![commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![tested with jest](https://img.shields.io/badge/tested_with-jest-99424f.svg)](https://github.com/facebook/jest)
[![auto release](https://img.shields.io/badge/release-auto.svg?colorA=888888&colorB=9B065A&label=auto)](https://github.com/intuit/auto)

## Installation

```sh
# with npm
npm install --save @rsql/ast

# with yarn
yarn add @rsql/ast
```

## API

#### `create<Node>(...args): Node`

```typescript
function createSelectorNode(selector): SelectorNode;
function createValueNode(value): ValueNode;
function createComparisionNode(selector, operator, value): ComparisionNode;
function createLogicNode(left, operator, right): LogicNode;
```

Creates new `Node`, can throw `InvalidArgumentError` for invalid arguments

#### `is<Node>(candidate): boolean`

```typescript
function isNode(candidate): boolean;
function isSelectorNode(candidate): boolean;
function isValueNode(candidate): boolean;
function isComparisionNode(candidate, operator?): boolean;
function isLogicNode(candidate, operator?): boolean;
function isExpressionNode(candidate): boolean;
```

Checks if `candidate` is an instance of `Node`
(actually it doesn't use `instanceof` operator as Nodes are simple objects - not instances of a class).
Additionally we can pass `operator` argument to the `isComparisionNode`
and `isLogicNode` to check if node uses given operator. It automatically maps operators to canonical
versions so for example `isComparisionNode(node, '=le=')` is same as `isComparisionNode(node, '<=')`.

#### `getSelector(comparision): string`

Returns selector for a given `ComparisionNode`.
Can throw `InvalidArgumentError` if passed argument is not a `ComparisionNode`.

#### `getValue(comparision): string | string[]`

Returns value for a given `ComparisionNode`.
Can throw `InvalidArgumentError` if passed argument is not a `ComparisionNode`.

#### `getSingleValue(comparision): string`

Returns single value for a given `ComparisionNode`.
Can throw `InvalidArgumentError` if passed argument is not a `ComparisionNode` or
contains array instead of string.

#### `getMultiValue(comparision): string[]`

Returns multi value for a given `ComparisionNode`.
Can throw `InvalidArgumentError` if passed argument is not a `ComparisionNode` or
contains string instead of array.

#### `ComparisionOperator`

```typescript
const EQ = "==";
const NEQ = "!=";
const LE = "<=";
const GE = ">=";
const LT = "<";
const GT = ">";
const IN = "=in=";
const OUT = "=out=";
const LE_VERBOSE = "=le=";
const GE_VERBOSE = "=ge=";
const LT_VERBOSE = "=lt=";
const GT_VERBOSE = "=gt=";

const CanonicalComparisionOperators = [EQ, NEQ, LE, GE, LT, GT, IN, OUT];
const VerboseComparisionOperators = [LE_VERBOSE, GE_VERBOSE, LT_VERBOSE, GT_VERBOSE];
const ComparisionOperators = [...CanonicalComparisionOperators, ...VerboseComparisionOperators];

function isComparisionOperator(candidate, operator?): boolean;
```

Defines built-in comparision operators and `isComparisionOperator` function which checks if given candidate
is a valid `ComparisionOperator`. Additionally you can pass `operator` argument which checks if
given candidate equals operator from a semantic perspective - so for example
`isComparisionOperator(candidate, LE)` will give you the same result as `isComparisionOperator(candidate, LE_VERBOSE)`

#### `LogicOperator`

```typescript
const AND = ";";
const OR = ",";
const AND_VERBOSE = "and";
const OR_VERBOSE = "or";

const CanonicalLogicOperators = [AND, OR];
const VerboseLogicOperators = [AND_VERBOSE, OR_VERBOSE];
const LogicOperators = [...CanonicalLogicOperators, ...VerboseLogicOperators];

function isLogicOperator(candidate, operator?): boolean;
```

Defines logic operators and `isLogicOperator` function which checks if given candidate
is a valid `LogicOperator`. Additionally you can pass `operator` argument which checks if
given candidate equals operator from a semantic perspective - so for example
`isLogicOperator(candidate, AND)` will give you the same result as `isLogicOperator(candidate, AND_VERBOSE)`

#### `ReservedChars`

```typescript
const ReservedChars = ['"', "'", "(", ")", ";", ",", "=", "!", "~", "<", ">", " ", "\n", "\t", "\r"];
```

Defines list of chars that are reserved in the RSQL.
Used internally to validate selectors, parse RSQL, and escape values.

## Types

#### `Node`

It's a base type for all node types. Defines an object which contains `type` property which defines
type of the node (instead of basing on the `instanceof` operator).

#### `SelectorNode`

Node which represents selector. Defines an object which, besides `Node` properties, contains
a `selector` property which stores selector.

#### `ValueNode`

Node which represents value. Defines an object which, besides `Node` properties, contains
a `value` property which stores value.

#### `BinaryNode`

It's a base type for nodes which consists of two operands and one operator.
Defines an object which, besides `Node` properties, contains `left` and `right` property which
stores operands and `operator` property which stores an operator.

#### `ComparisionNode`

Node which represents comparision expression. It's a constrained version of the `BinaryNode` where
`left` has to be `SelectorNode`, `right` has to be a `ValueNode` and `operator` has to be a
`ComparisionOperator`.

#### `LogicNode`

Node which represents logic expression. It's a constrained version of the `BinaryNode` where
`left` and `right` has to be an `ExpressionNode` and `operator` has to be a `LogicOperator`.

#### `ExpressionNode`

It's a type union between `ComparisionNode` and `LogicNode`.

#### `ComparisionOperator`

It's a string literal type which defines built-in comparision operators.

#### `LogicOperator`

It's a string literal type which defines built-in logic operators.

## License

MIT
