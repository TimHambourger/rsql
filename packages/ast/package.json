{
  "name": "@rsql/ast",
  "description": "RSQL AST",
  "repository": "https://github.com/piotr-oles/rsql.git",
  "license": "MIT",
  "author": "Piotr Ole≈õ <piotrek.oles@gmail.com>",
  "version": "0.0.0",
  "sideEffects": false,
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "scripts": {
    "clean": "rm -rf ./lib && rm -f tsconfig.tsbuildinfo",
    "compile": "tsc -b tsconfig.json",
    "dev": "yarn compile --watch --preserveWatchOutput",
    "build": "yarn clean && yarn compile"
  },
  "engines": {
    "node": ">= 8"
  },
  "dependencies": {
    "@rsql/definitions": "0.0.0"
  },
  "jest": {
    "preset": "ts-jest",
    "testEnvironment": "node"
  },
  "publishConfig": {
    "access": "public"
  }
}
