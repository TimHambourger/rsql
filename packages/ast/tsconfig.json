{
  "extends": "../../tsconfig",
  "compilerOptions": {
    "composite": true,
    "rootDir": "./src",
    "outDir": "./lib"
  },
  "exclude": ["node_modules", "lib", "dist"]
}
