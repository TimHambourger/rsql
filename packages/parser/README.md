# @rsql/parser

RSQL parser for Node.js and Browsers

⚠️ WARNING: This package is still in development - API may break compatibility without upgrading major version! ⚠️

[![npm](https://img.shields.io/npm/v/@rsql/parser)](https://www.npmjs.com/package/@rsql/parser)
[![lerna](https://img.shields.io/badge/maintained%20with-lerna-cc00ff.svg)](https://lerna.js.org/)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier)
[![commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![tested with jest](https://img.shields.io/badge/tested_with-jest-99424f.svg)](https://github.com/facebook/jest)
[![auto release](https://img.shields.io/badge/release-auto.svg?colorA=888888&colorB=9B065A&label=auto)](https://github.com/intuit/auto)

## Installation

```sh
# with npm
npm install --save @rsql/parser

# with yarn
yarn add @rsql/parser
```

## API

#### `parse(source: string): ExpressionNode`

Parses RSQL string and returns Abstract Syntax Tree. It can throw the following errors:

- `TypeError` - in the case of invalid argument type passed to the `parse` function
- `SyntaxError` - in the case of any problems encountered during parsing

## License

MIT
