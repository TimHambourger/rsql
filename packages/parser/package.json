{
  "name": "@rsql/parser",
  "description": "RSQL Parser",
  "repository": "https://github.com/piotr-oles/rsql.git",
  "license": "MIT",
  "author": "Piotr Ole≈õ <piotrek.oles@gmail.com>",
  "version": "0.0.0",
  "sideEffects": false,
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "module": "dist/index.esm.js",
  "browser": "dist/index.umd.js",
  "files": [
    "dist"
  ],
  "keywords": [
    "rsql",
    "fiql",
    "parser"
  ],
  "scripts": {
    "clean": "rm -rf ./lib && rm -f tsconfig.tsbuildinfo && rm -rf ./dist",
    "compile": "tsc -b tsconfig.json && rollup -c && dts-bundle --name \"@rsql/parser\" --main lib/index.d.ts --out ../dist/index.d.ts",
    "build": "yarn clean && yarn compile"
  },
  "engines": {
    "node": ">= 8"
  },
  "dependencies": {
    "@rsql/ast": "0.0.0"
  },
  "jest": {
    "preset": "ts-jest",
    "testEnvironment": "node"
  },
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  }
}
