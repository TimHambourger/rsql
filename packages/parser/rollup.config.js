import resolve from "@rollup/plugin-node-resolve";
import commonjs from "@rollup/plugin-commonjs";
import pkg from "./package.json";

export default [
  // browser-friendly UMD build
  {
    input: "lib/index.js",
    output: {
      name: "RSQLParser",
      file: pkg.browser,
      format: "umd",
    },
    plugins: [resolve(), commonjs()],
  },
  // CommonJS (for Node) and ES module (for bundlers) build.
  {
    input: "lib/index.js",
    external: Object.keys(pkg.dependencies),
    output: [
      { file: pkg.main, format: "cjs" },
      { file: pkg.module, format: "es" },
    ],
  },
];
