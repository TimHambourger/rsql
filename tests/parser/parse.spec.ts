import { isComparisionExpressionNode, isLogicExpressionNode, isSelectorNode, isValueNode } from "@rsql/ast";
import { parse } from "@rsql/parser";

function assert(condition: unknown): asserts condition {
  expect(condition).toBeTruthy();
}

describe("parse", () => {
  it.each([
    ["==", "=="],
    ["!=", "!="],
    ["<=", "<="],
    [">=", ">="],
    ["<", "<"],
    [">", ">"],
    ["=in=", "=in="],
    ["=out=", "=out="],
    ["=le=", "<="],
    ["=ge=", ">="],
    ["=lt=", "<"],
    ["=gt=", ">"],
  ])("parses comparision expression for operator %p", (operator, canonicalOperator) => {
    const rsql = `selector${operator}value`;
    const comparision = parse(rsql);

    assert(isComparisionExpressionNode(comparision));
    assert(isSelectorNode(comparision.left));
    assert(isValueNode(comparision.right));

    expect(comparision.operator).toEqual(canonicalOperator);
    expect(comparision.left.selector).toEqual("selector");
    expect(comparision.right.value).toEqual("value");
  });

  it('throws exception for comparision operator typo "="', () => {
    expect(() => parse("selector=value")).toThrowError("Unexpected character '=' at position 9 in \"selector=value\"");
  });

  it.each(["allons-y", "l00k.dot.path", "look/XML/path", "n:look/n:xml", "path.to::Ref", "$doll_r.way"])(
    'parses selector "%p"',
    (selector) => {
      const rsql = `${selector}==value`;
      const comparision = parse(rsql);

      assert(isComparisionExpressionNode(comparision));
      assert(isSelectorNode(comparision.left));
      assert(isValueNode(comparision.right));

      expect(comparision.operator).toEqual("==");
      expect(comparision.left.selector).toEqual(selector);
      expect(comparision.right.value).toEqual("value");
    }
  );

  it.each([
    ['ill"', `Unclosed quote '"' at position 4 in "ill"==value"`],
    ["ill'", `Unclosed quote ''' at position 4 in "ill'==value"`],
    ["ill(", `Unexpected character '(' at position 4 in "ill(==value"`],
    ["ill)", `Unexpected character ')' at position 4 in "ill)==value"`],
    ["ill;", `Unexpected character ';' at position 4 in "ill;==value"`],
    ["ill,", `Unexpected character ',' at position 4 in "ill,==value"`],
    ["ill=", `Unexpected character '=' at position 6 in "ill===value"`],
    ["ill<", `Unexpected character '=' at position 6 in "ill<==value`],
    ["ill>", `Unexpected character '=' at position 6 in "ill>==value`],
    ["ill!", `Unexpected character '=' at position 6 in "ill!==value`],
    ["ill~", `Unexpected character '~' at position 4 in "ill~==value`],
    ['ill"ness', `Unclosed quote '"' at position 4 in "ill"ness==value`],
    ["ill'ness", `Unclosed quote ''' at position 4 in "ill'ness==value`],
    ["ill(ness", `Unexpected character '(' at position 4 in "ill(ness==value"`],
    ["ill)ness", `Unexpected character ')' at position 4 in "ill)ness==value"`],
    ["ill;ness", `Unexpected character ';' at position 4 in "ill;ness==value"`],
    ["ill,ness", `Unexpected character ',' at position 4 in "ill,ness==value"`],
    ["ill=ness", `Unexpected character '=' at position 4 in "ill=ness==value`],
    ["ill<ness", `Unexpected character '==' at position 9 in "ill<ness==value"`],
    ["ill>ness", `Unexpected character '==' at position 9 in "ill>ness==value"`],
    ["ill!ness", `Unexpected character '!' at position 4 in "ill!ness==value`],
    ["ill~ness", `Unexpected character '~' at position 4 in "ill~ness==value`],
  ])('throws error for selector with reserved char "%p"', (selector, error) => {
    expect(() => parse(`${selector}==value`)).toThrowError(error);
  });

  it("throws exception for empty selector", () => {
    expect(() => parse("==value")).toThrowError(`Unexpected character '==' at position 1 in "==value"`);
  });

  it.each(["«Allons-y»", "h@llo", "*star*", "čes*ký", "42", "0.15", "3:15"])('parses unquoted value "%p"', (value) => {
    const rsql = `selector==${value}`;
    const comparision = parse(rsql);

    assert(isComparisionExpressionNode(comparision));
    assert(isSelectorNode(comparision.left));
    assert(isValueNode(comparision.right));

    expect(comparision.operator).toEqual("==");
    expect(comparision.left.selector).toEqual("selector");
    expect(comparision.right.value).toEqual(value);
  });

  it.each([
    ['ill"', `Unclosed quote '"' at position 14 in "selector==ill""`],
    ["ill'", `Unclosed quote ''' at position 14 in "selector==ill'"`],
    ["ill(", `Unexpected character '(' at position 14 in "selector==ill("`],
    ["ill)", `Unexpected character ')' at position 14 in "selector==ill)"`],
    ["ill;", `Unexpected end in "selector==ill;"`],
    ["ill,", `Unexpected end in "selector==ill,"`],
    ["ill=", `Unexpected character '=' at position 14 in "selector==ill="`],
    ["ill<", `Unexpected character '<' at position 14 in "selector==ill<"`],
    ["ill>", `Unexpected character '>' at position 14 in "selector==ill>"`],
    ["ill!", `Unexpected character '!' at position 14 in "selector==ill!"`],
    ["ill~", `Unexpected character '~' at position 14 in "selector==ill~"`],
    ['ill"ness', `Unclosed quote '"' at position 14 in "selector==ill"ness"`],
    ["ill'ness", `Unclosed quote ''' at position 14 in "selector==ill'ness"`],
    ["ill(ness", `Unexpected character '(' at position 14 in "selector==ill(ness"`],
    ["ill)ness", `Unexpected character ')' at position 14 in "selector==ill)ness"`],
    ["ill;ness", `Unexpected end in "selector==ill;ness"`],
    ["ill,ness", `Unexpected end in "selector==ill,ness"`],
    ["ill=ness", `Unexpected character '=' at position 14 in "selector==ill=ness"`],
    ["ill<ness", `Unexpected character '<' at position 14 in "selector==ill<ness"`],
    ["ill>ness", `Unexpected character '>' at position 14 in "selector==ill>ness"`],
    ["ill!ness", `Unexpected character '!' at position 14 in "selector==ill!ness"`],
    ["ill~ness", `Unexpected character '~' at position 14 in "selector==ill~ness"`],
  ])('throws error for unquoted value with reserved char "%p"', (value, error) => {
    expect(() => parse(`selector==${value}`)).toThrowError(error);
  });

  it("throws an error for empty selector", () => {
    expect(() => parse("==value")).toThrowError(`Unexpected character '==' at position 1 in "==value"`);
  });

  it.each(['"hi there!"', "'Pěkný den!'", '"Flynn\'s *"', "\"o)'O'(o\"", '"6*7=42"', '""'])(
    'parses quoted value with any chars "%p"',
    (value) => {
      const rsql = `selector==${value}`;
      const comparision = parse(rsql);

      assert(isComparisionExpressionNode(comparision));
      assert(isSelectorNode(comparision.left));
      assert(isValueNode(comparision.right));

      expect(comparision.operator).toEqual("==");
      expect(comparision.left.selector).toEqual("selector");
      expect(comparision.right.value).toEqual(value.slice(1, -1));
    }
  );

  it.each([
    "'10\\' 15\"'",
    "'10\\' 15\\\"'",
    "'w\\\\ \\'Flyn\\n\\''",
    "'\\\\(^_^)/'",
    "'\\\\'",
    '"10\' 15\\""',
    '"10\\\' 15\\""',
    '"w\\\\ \\"Flyn\\n\\""',
    '"\\\\(^_^)/"',
    '"\\\\"',
  ])('parses escaped quoted value "%p"', (value) => {
    const rsql = `selector==${value}`;
    const comparision = parse(rsql);

    assert(isComparisionExpressionNode(comparision));
    assert(isSelectorNode(comparision.left));
    assert(isValueNode(comparision.right));

    expect(comparision.operator).toEqual("==");
    expect(comparision.left.selector).toEqual("selector");
    expect(comparision.right.value).toEqual(value.slice(1, -1));
  });

  it.each([
    [
      ["chunky", "bacon", '"ftw!"'],
      ["chunky", "bacon", "ftw!"],
    ],
    [
      ["'hi!'", '"how\'re you?"'],
      ["hi!", "how're you?"],
    ],
    [["meh"], ["meh"]],
    [['")o("'], [")o("]],
  ])('parses values group "%p"', (values, expectedValues) => {
    const rsql = `selector=in=(${values.join(",")})`;
    const comparision = parse(rsql);

    assert(isComparisionExpressionNode(comparision));
    assert(isSelectorNode(comparision.left));
    assert(isValueNode(comparision.right));

    expect(comparision.operator).toEqual("=in=");
    expect(comparision.left.selector).toEqual("selector");
    expect(comparision.right.value).toEqual(expectedValues);
  });

  it("throws an error for empty value", () => {
    expect(() => parse("selector=in=()")).toThrowError(`Unexpected character ')' at position 14 in "selector=in=()"`);
  });

  it.each([
    [",", ","],
    [";", ";"],
    [" and ", ";"],
    [" or ", ","],
  ])('parses logical operator "%p"', (operator, canonicalOperator) => {
    const rsql = `selector1==value1${operator}selector2!=value2`;
    const logic = parse(rsql);

    assert(isLogicExpressionNode(logic));
    assert(isComparisionExpressionNode(logic.left));
    expect(logic.operator).toEqual(canonicalOperator);
    assert(isComparisionExpressionNode(logic.right));

    // left comparision
    assert(isSelectorNode(logic.left.left));
    expect(logic.left.left.selector).toEqual("selector1");
    expect(logic.left.operator).toEqual("==");
    assert(isValueNode(logic.left.right));
    expect(logic.left.right.value).toEqual("value1");

    // right comparision
    assert(isSelectorNode(logic.right.left));
    expect(logic.right.left.selector).toEqual("selector2");
    expect(logic.right.operator).toEqual("!=");
    assert(isValueNode(logic.right.right));
    expect(logic.right.right.value).toEqual("value2");
  });

  it.each(["s0==v0;s1==v1;s2==v2", "s0==v0,s1=out=(v10,v11),s2==v2", "s0==v0,s1==v1;s2==v2,s3==v3"])(
    'parses query with operator precedence "%p"',
    (rsql) => {
      const ast = parse(rsql);

      expect(ast).toMatchSnapshot();
    }
  );

  it.each([
    "(s0==a0,s1==a1);s2==a2",
    "(s0==a0,s1=out=(a10,a11));s2==a2,s3==a3",
    "((s0==a0,s1==a1);s2==a2,s3==a3);s4==a4",
    "(s0==a0)",
    "((s0==a0));s1==a1",
  ])('parses query with parenthesis "%p"', (rsql) => {
    const ast = parse(rsql);

    expect(ast).toMatchSnapshot();
  });

  it.each([
    ["(s0==a0;s1!=a1", 'Unexpected end in "(s0==a0;s1!=a1". Did you forget to close parenthesis at position 1?'],
    ["s0==a0)", "Unexpected character ')' at position 7 in \"s0==a0)\""],
    [
      "s0==a;(s1=in=(b,c),s2!=d",
      'Unexpected end in "s0==a;(s1=in=(b,c),s2!=d". Did you forget to close parenthesis at position 5?',
    ],
  ])('throws an error for unclosed parenthesis "%p"', (rsql, error) => {
    expect(() => parse(rsql)).toThrowError(error);
  });
});
