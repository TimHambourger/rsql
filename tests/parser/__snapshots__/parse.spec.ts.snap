// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse parses query with operator precedence ""s0==v0,s1==v1;s2==v2,s3==v3"" 1`] = `
Object {
  "left": Object {
    "left": Object {
      "left": Object {
        "selector": "s0",
        "type": "SELECTOR",
      },
      "operator": "==",
      "right": Object {
        "type": "VALUE",
        "value": "v0",
      },
      "type": "COMPARISON",
    },
    "operator": ",",
    "right": Object {
      "left": Object {
        "left": Object {
          "selector": "s1",
          "type": "SELECTOR",
        },
        "operator": "==",
        "right": Object {
          "type": "VALUE",
          "value": "v1",
        },
        "type": "COMPARISON",
      },
      "operator": ";",
      "right": Object {
        "left": Object {
          "selector": "s2",
          "type": "SELECTOR",
        },
        "operator": "==",
        "right": Object {
          "type": "VALUE",
          "value": "v2",
        },
        "type": "COMPARISON",
      },
      "type": "LOGIC",
    },
    "type": "LOGIC",
  },
  "operator": ",",
  "right": Object {
    "left": Object {
      "selector": "s3",
      "type": "SELECTOR",
    },
    "operator": "==",
    "right": Object {
      "type": "VALUE",
      "value": "v3",
    },
    "type": "COMPARISON",
  },
  "type": "LOGIC",
}
`;

exports[`parse parses query with operator precedence ""s0==v0,s1=out=(v10,v11),s2==v2"" 1`] = `
Object {
  "left": Object {
    "left": Object {
      "left": Object {
        "selector": "s0",
        "type": "SELECTOR",
      },
      "operator": "==",
      "right": Object {
        "type": "VALUE",
        "value": "v0",
      },
      "type": "COMPARISON",
    },
    "operator": ",",
    "right": Object {
      "left": Object {
        "selector": "s1",
        "type": "SELECTOR",
      },
      "operator": "=out=",
      "right": Object {
        "type": "VALUE",
        "value": Array [
          "v10",
          "v11",
        ],
      },
      "type": "COMPARISON",
    },
    "type": "LOGIC",
  },
  "operator": ",",
  "right": Object {
    "left": Object {
      "selector": "s2",
      "type": "SELECTOR",
    },
    "operator": "==",
    "right": Object {
      "type": "VALUE",
      "value": "v2",
    },
    "type": "COMPARISON",
  },
  "type": "LOGIC",
}
`;

exports[`parse parses query with operator precedence ""s0==v0;s1==v1;s2==v2"" 1`] = `
Object {
  "left": Object {
    "left": Object {
      "left": Object {
        "selector": "s0",
        "type": "SELECTOR",
      },
      "operator": "==",
      "right": Object {
        "type": "VALUE",
        "value": "v0",
      },
      "type": "COMPARISON",
    },
    "operator": ";",
    "right": Object {
      "left": Object {
        "selector": "s1",
        "type": "SELECTOR",
      },
      "operator": "==",
      "right": Object {
        "type": "VALUE",
        "value": "v1",
      },
      "type": "COMPARISON",
    },
    "type": "LOGIC",
  },
  "operator": ";",
  "right": Object {
    "left": Object {
      "selector": "s2",
      "type": "SELECTOR",
    },
    "operator": "==",
    "right": Object {
      "type": "VALUE",
      "value": "v2",
    },
    "type": "COMPARISON",
  },
  "type": "LOGIC",
}
`;

exports[`parse parses query with parenthesis ""((s0==a0));s1==a1"" 1`] = `
Object {
  "left": Object {
    "left": Object {
      "selector": "s0",
      "type": "SELECTOR",
    },
    "operator": "==",
    "right": Object {
      "type": "VALUE",
      "value": "a0",
    },
    "type": "COMPARISON",
  },
  "operator": ";",
  "right": Object {
    "left": Object {
      "selector": "s1",
      "type": "SELECTOR",
    },
    "operator": "==",
    "right": Object {
      "type": "VALUE",
      "value": "a1",
    },
    "type": "COMPARISON",
  },
  "type": "LOGIC",
}
`;

exports[`parse parses query with parenthesis ""((s0==a0,s1==a1);s2==a2,s3==a3);s4==a4"" 1`] = `
Object {
  "left": Object {
    "left": Object {
      "left": Object {
        "left": Object {
          "left": Object {
            "selector": "s0",
            "type": "SELECTOR",
          },
          "operator": "==",
          "right": Object {
            "type": "VALUE",
            "value": "a0",
          },
          "type": "COMPARISON",
        },
        "operator": ",",
        "right": Object {
          "left": Object {
            "selector": "s1",
            "type": "SELECTOR",
          },
          "operator": "==",
          "right": Object {
            "type": "VALUE",
            "value": "a1",
          },
          "type": "COMPARISON",
        },
        "type": "LOGIC",
      },
      "operator": ";",
      "right": Object {
        "left": Object {
          "selector": "s2",
          "type": "SELECTOR",
        },
        "operator": "==",
        "right": Object {
          "type": "VALUE",
          "value": "a2",
        },
        "type": "COMPARISON",
      },
      "type": "LOGIC",
    },
    "operator": ",",
    "right": Object {
      "left": Object {
        "selector": "s3",
        "type": "SELECTOR",
      },
      "operator": "==",
      "right": Object {
        "type": "VALUE",
        "value": "a3",
      },
      "type": "COMPARISON",
    },
    "type": "LOGIC",
  },
  "operator": ";",
  "right": Object {
    "left": Object {
      "selector": "s4",
      "type": "SELECTOR",
    },
    "operator": "==",
    "right": Object {
      "type": "VALUE",
      "value": "a4",
    },
    "type": "COMPARISON",
  },
  "type": "LOGIC",
}
`;

exports[`parse parses query with parenthesis ""(s0==a0)"" 1`] = `
Object {
  "left": Object {
    "selector": "s0",
    "type": "SELECTOR",
  },
  "operator": "==",
  "right": Object {
    "type": "VALUE",
    "value": "a0",
  },
  "type": "COMPARISON",
}
`;

exports[`parse parses query with parenthesis ""(s0==a0,s1==a1);s2==a2"" 1`] = `
Object {
  "left": Object {
    "left": Object {
      "left": Object {
        "selector": "s0",
        "type": "SELECTOR",
      },
      "operator": "==",
      "right": Object {
        "type": "VALUE",
        "value": "a0",
      },
      "type": "COMPARISON",
    },
    "operator": ",",
    "right": Object {
      "left": Object {
        "selector": "s1",
        "type": "SELECTOR",
      },
      "operator": "==",
      "right": Object {
        "type": "VALUE",
        "value": "a1",
      },
      "type": "COMPARISON",
    },
    "type": "LOGIC",
  },
  "operator": ";",
  "right": Object {
    "left": Object {
      "selector": "s2",
      "type": "SELECTOR",
    },
    "operator": "==",
    "right": Object {
      "type": "VALUE",
      "value": "a2",
    },
    "type": "COMPARISON",
  },
  "type": "LOGIC",
}
`;

exports[`parse parses query with parenthesis ""(s0==a0,s1=out=(a10,a11));s2==a2,s3==a3"" 1`] = `
Object {
  "left": Object {
    "left": Object {
      "left": Object {
        "left": Object {
          "selector": "s0",
          "type": "SELECTOR",
        },
        "operator": "==",
        "right": Object {
          "type": "VALUE",
          "value": "a0",
        },
        "type": "COMPARISON",
      },
      "operator": ",",
      "right": Object {
        "left": Object {
          "selector": "s1",
          "type": "SELECTOR",
        },
        "operator": "=out=",
        "right": Object {
          "type": "VALUE",
          "value": Array [
            "a10",
            "a11",
          ],
        },
        "type": "COMPARISON",
      },
      "type": "LOGIC",
    },
    "operator": ";",
    "right": Object {
      "left": Object {
        "selector": "s2",
        "type": "SELECTOR",
      },
      "operator": "==",
      "right": Object {
        "type": "VALUE",
        "value": "a2",
      },
      "type": "COMPARISON",
    },
    "type": "LOGIC",
  },
  "operator": ",",
  "right": Object {
    "left": Object {
      "selector": "s3",
      "type": "SELECTOR",
    },
    "operator": "==",
    "right": Object {
      "type": "VALUE",
      "value": "a3",
    },
    "type": "COMPARISON",
  },
  "type": "LOGIC",
}
`;
